### PR TITLE
:bug:  create empty pull secret in the operator ns

### PR DESCRIPTION
**Pull request by zhiweiyin318**
[comment]: # ( Copyright Contributors to the Open Cluster Management project )

## Related issue(s)

https://github.com/open-cluster-management-io/clusteradm/issues/433

Fixes #

### CODE REVIEW

File: pkg/cmd/init/exec.go [hunk 1]
```diff
@@ -219,7 +219,13 @@ func (o *Options) run() error {
 			}
 		}
 
-		//if service-account wait for the sa secret
+		//  apply an empty image pull secret if the image pull secret is not found in the ns.
+		err = helpers.ApplyImagePullSecret(context.TODO(), "open-cluster-management", kubeClient)
```
Comment (zhujian7) [inline pkg/cmd/init/exec.go:223]: if it is dry run, will we create the secret?

File: pkg/cmd/join/exec.go [hunk 2]
```diff
@@ -330,6 +330,12 @@ func (o *Options) run() error {
 		return err
 	}
 
+	//  apply an empty image pull secret if the image pull secret is not found in the ns.
+	err = helpers.ApplyImagePullSecret(context.TODO(), "open-cluster-management", kubeClient)
```
Comment (zhujian7) [inline pkg/cmd/join/exec.go:334]: please also consider dry-run, and I am not sure if this should be put into the above `applyKlusterlet` func, is this secret used by the klusterlet-operator? if it is, I think we should put it into the `applyKlusterlet` func, and use the constant `OperatorNamesapce` to replace the "open-cluster-management"

File: pkg/cmd/init/exec.go (hunk 1 above)
Comment (zhiweiyin318) [reply pkg/cmd/init/exec.go]: I think we need to create the secret . have fixed.

File: pkg/cmd/join/exec.go (hunk 2 above)
Comment (zhiweiyin318) [reply pkg/cmd/join/exec.go]: There is no image pull secret in the operator deployment or sa. so no need to consider the order.

File: pkg/cmd/init/scenario/init/image-pull-secret.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+apiVersion: v1
+kind: Secret
+metadata:
+  name: open-cluster-management-image-pull-credentials
+  namespace: open-cluster-management
+  labels:
+    app: cluster-manager
+type: kubernetes.io/dockerconfigjson
+data:
+  .dockerconfigjson: e30K  # "{}" empty value here
```
[the remark this answers is not in the document]
Comment (zhujian7) [reply pkg/cmd/init/scenario/init/image-pull-secret.yaml:11]: oh, it is secret, so should be base64ed.

**Review by qiujian16 — COMMENTED** [review]
ideally, should have a flag to input docker creds, but we could do it in the future.

File: pkg/helpers/client.go [hunk 4]
```diff
@@ -38,6 +38,8 @@ const (
 	BootstrapToken      TokenType = "bootstrap-token"
 	ServiceAccountToken TokenType = "service-account-token"
 	UnknownToken        TokenType = "unknown-token"
+
+	ImagePullSecret = "open-cluster-management-image-pull-credentials"
```
[the remark this answers is not in the document]
Comment (zhiweiyin318) [reply pkg/helpers/client.go:42]: this is used in e2e test, let me move it to e2e part.

File: pkg/cmd/init/cmd.go [hunk 5]
```diff
@@ -59,10 +59,12 @@ func NewCmd(clusteradmFlags *genericclioptionsclusteradm.ClusteradmFlags, stream
 	cmd.Flags().StringToStringVar(&o.resourceRequests, "resource-requests", nil, "the resource requests of all the containers managed by the cluster manager and the cluster manager operator, for example: cpu=500m,memory=500Mi")
 	cmd.Flags().BoolVar(&o.createNamespace, "create-namespace", true, "If true, create open-cluster-management namespace, otherwise use existing one")
 
-	//clusterManagetSet contains the flags for deploy cluster-manager
+	// clusterManagetSet contains the flags for deploy cluster-manager
 	clusterManagerSet := pflag.NewFlagSet("clusterManagerSet", pflag.ExitOnError)
 	cmd.Flags().StringVar(&o.registry, "image-registry", "quay.io/open-cluster-management",
 		"The name of the image registry serving OCM images, which will be applied to all the deploying OCM components.")
+	cmd.Flags().StringVar(&o.imagePullCred, "image-pull-credential", "",
```
Comment (qiujian16) [inline pkg/cmd/init/cmd.go:66]: actually, I think we should set a file instead of a base64 encoded string, it is not user friendly.

**Comment by zhiweiyin318** [comment]
@qiujian16 @zhujian7  please take another look.
1. changed to `image-pull-credential-file`.
2. add the pull secret to operator deployments too since the user can configure it via flag.

**Pull request merged**